### PR TITLE
[graphql][2/n] Refactor convert_to_transaction_block

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Continuing off https://github.com/MystenLabs/sui/pull/13714. 
It looks like the effects field of TransactionBlock have some heavy overlaps with other top-level fields of TransactionBlock. We already have the data needed to resolve every field on effects: TransactionBlockEffects. In fact, if a caller specifies both tx.effects.epoch and tx.expiration, we make redundant calls for sui_system_state and fetch_protocol_config. It's also a little strange to me that we pass tx_effects to TransactionBlockEffects for status and errors. Wondering if it's worth it right now to modify schema for TransactionBlock and TransactionBlockEffects? We could remove the redundant digest, lift status and errors to top-level, etc.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/types/transaction_block.rs:85]: Don't unwrap here, use Result instead.

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/types/transaction_block.rs:30]: Why return Result here?

File: crates/sui-graphql-rpc/src/server/sui_sdk_data_provider.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/server/sui_sdk_data_provider.rs:592]: @wlmyng your question about redundancies in the schema got me curious, so I had a look here.

This digest is not redundant, it's just filled in incorrectly here.  This should be the **effects digest**, not the transaction digest as it is now (i.e. this is a bug).

We shouldn't change the schema here -- in the protocol, the transaction and its effects are distinct structures, with distinct IDs (digests).

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/transaction_block.rs:34]: Porting over the comment from above -- this shouldn't be the tx digest, it should be the effects digest.

**Review by amnn — COMMENTED** [review]
Supportive of the change, but note the issue you've uncovered around digests, which needs fixing.

Generally, we need better test coverage here (and in the PR below in the stack) -- it would have made the issue with the repeated digest more obvious, and it's also incredibly helpful to make sure we preserve behaviour across refactors like this.

I thought we (@stefan-mysten ?) were looking into having tests that snapshot queries against a static DB -- what's the status of that?

File: crates/sui-graphql-rpc/src/server/sui_sdk_data_provider.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/server/sui_sdk_data_provider.rs:592]: I see, that makes a lot of sense 😅

**Comment by stefan-mysten** [comment]
>  thought we (@stefan-mysten ?) were looking into having tests that snapshot queries against a static DB -- what's the status of that?
I have not worked on this yet, and I don't think anyone else started on it either.

**Comment by amnn** [comment]
Okay good to know.  As we start doing larger refactorings, this work is going to be more crucial and worth prioritising, cc @oxade .

**Comment by wlmyng** [comment]
In https://github.com/MystenLabs/sui/pull/13714#issuecomment-1713219691 I proposed that any object with fields that cannot be immediately returned should be resolved in Object resolvers. After working on some more refactoring, I think we're better off by doing direct conversions for simple fields and porting just edge resolution and complex computations.

Some justifications:
1. Creating intermediary structures just for the sake of GraphQL conversion leads to unnecessary complexity. GasCostSummary from the linked PR is a good example. We have From<json_rpc::GasCostSummary> for graphql::GasCostSummary and an impl Object block just to map u64 to BigInt. I think it's just easier to do the conversions directly.
2. In this PR's earlier commits, I found myself wrapping TransactionBlock over sui_json_rpc_types::SuiJsonTransactionBlockResponse. Doing data conversions upfront helps decouple us from external crates
3. We can leverage SimpleObject + ComplexObject. Most of TransactionBlock's fields are ready to go, and we just defer resolution of the expiration field to the ComplexObject block. Note that even though we are instantiating Effects, which itself is a ComplexObject, we are not populating all fields, so we still follow graphql's lazy resolution practice

File: crates/sui-graphql-rpc/src/context_data/sui_sdk_data_provider.rs
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/sui_sdk_data_provider.rs:408]: I think here we want `rolling_gas_summary.clone()`, might make it cleaner.

Comment (stefan-mysten) [reply crates/sui-graphql-rpc/src/context_data/sui_sdk_data_provider.rs:408]: Or pass a reference instead of the giving up ownership?

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/types/transaction_block.rs:32]: Consider adding an import of `SuiTransactionBlockResponse` instead of the qualified version here and on the next line.

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/types/transaction_block.rs:33]: Ditto

**Review by stefan-mysten — APPROVED** [review]

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/transaction_block.rs:93]: This is the wrong digest, and unfortunately I believe the JSON RPC doesn't store the effects digest, so you may not be able to access it via this data provider -- so I would suggest leaving this field off for now.

I have left a [comment on the data layer schema](https://github.com/MystenLabs/sui/pull/13574/files#r1324231969) enquiring whether we should support effect digests there, or remove support for them on the GraphQL side.  Let's wait to here what they say.

P.S., as per my comment here: https://github.com/MystenLabs/sui/pull/13660#pullrequestreview-1615050100 we shouldn't have a dedicated scalar for `TransactionDigest`.  It should just be a `String` based on our initial design discussions where we decided to limit our use of custom scalars, to reduce the cost for adopting this schema on the client (each custom scalar must be backed by a custom serializer/deserializer on the client).  In this case it's even adding to the confusion, because this isn't a `TransactionDigest`, it's an `EffectsDigest`.  Fixing this doesn't need to block this PR, but could you handle it in a follow-up @wlmyng ?

**Review by amnn — COMMENTED** [review]
It would be good if we could address the effects digest issue before landing this PR (i.e. just get rid of the effects digest), but it also wasn't introduced in this PR, so we don't need to block, if you want to address that in a follow-up.  Same goes for the comment about the `TransactionDigest` scalar.

**Comment by amnn** [comment]
Overall, I'm really liking these refactors :) Thanks for putting `Object` and `ComplexObject` into practice for us!

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/transaction_block.rs:57]: Yeah I think this makes sense

File: crates/sui-graphql-rpc/src/types/transaction_block.rs
Comment (amnn) [reply crates/sui-graphql-rpc/src/types/transaction_block.rs:93]: The way things are leaning so far is towards not having to refer to effects digest at all, so that makes things simpler!  For reference, I've updated the draft schema in https://github.com/MystenLabs/sui/pull/13765, to give an idea of what that means for us (still need to leave some time in case other folks have thoughts to share on this though, before locking that change in).

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/transaction_block.rs:93]: Ah missed your latest comment

**Pull request merged**
